solidity/dashboard/src/sagas/coverage-pool.js

```diff
@@ -1,27 +1,182 @@
+import {
+  identifyTaskByAddress,
+  logErrorAndThrow,
+  logError,
+  submitButtonHelper,
+} from "./utils"
+import { Keep } from "../contracts"
+import { add, sub } from "../utils/arithmetics.utils"
+import { isSameEthAddress } from "../utils/general.utils"
+import { ZERO_ADDRESS } from "../utils/ethereum.utils"
+import { sendTransaction } from "./web3"
 
 function* fetchTvl() {
   try {
     yield put(fetchTvlStart())
-    const tvl = yield call()
+    const tvl = yield call(Keep.coveragePoolV1.totalValueLocked)
     yield put(fetchTvlSuccess(tvl))
   } catch (error) {
-    yield* logErrorAndThrow(COVERAGE_POOL_FETCH_TVL_ERROR, error)
+    yield* logError(COVERAGE_POOL_FETCH_TVL_ERROR, error)
   }
 }
 
 export function* watchFetchTvl() {
+  yield takeLatest(COVERAGE_POOL_FETCH_TVL_REQUEST, fetchTvl)
+}
+
+function* fetchCovPoolData(action) {
+  const { address } = action.payload
+  try {
+    yield put(fetchCovPoolDataStart())
+
+    const balanceOf = yield call(Keep.coveragePoolV1.covBalanceOf, address)
+    const totalSupply = yield call(Keep.coveragePoolV1.covTotalSupply)
+    const shareOfPool = yield call(
+      Keep.coveragePoolV1.shareOfPool,
+      totalSupply,
+      balanceOf
+    )
+    const estimatedKeepBalance = yield call(
+      Keep.coveragePoolV1.estimatedCollateralTokenBalance,
+      shareOfPool
+    )
+
+    const estimatedRewards = yield call(
+      Keep.coveragePoolV1.estimatedRewards,
+      shareOfPool,
+      estimatedKeepBalance
+    )
+
+    yield put(
+      fetchCovPoolDataSuccess({
+        shareOfPool,
+        covBalance: balanceOf,
+        covTotalSupply: totalSupply,
+        estimatedRewards,
+        estimatedKeepBalance,
+      })
+    )
+  } catch (error) {
+    yield* logErrorAndThrow(COVERAGE_POOL_FETCH_COV_POOL_DATA_ERROR, error)
+  }
+}
+
+export function* watchFetchCovPoolData() {
   yield takeOnlyOnce(
-    COVERAGE_POOL_FETCH_TVL_REQUEST,
+    COVERAGE_POOL_FETCH_COV_POOL_DATA_REQUEST,
     identifyTaskByAddress,
-    fetchTvl
+    fetchCovPoolData
+  )
+}
+
+export function* subscribeToCovTokenTransferEvent() {
+  const requestChan = yield actionChannel(
+    COVERAGE_POOL_COV_TOKEN_TRANSFER_EVENT_EMITTED
   )
+
+  while (true) {
+    const {
+      payload: { event },
+    } = yield take(requestChan)
+    const {
+      returnValues: { from, to, value },
+    } = event
+    const { covTotalSupply, covBalance } = yield select(
+      (state) => state.coveragePool
+    )
+
+    const address = yield select((state) => state.app.address)
+    let updatedCovTotalSupply = 0
+    if (isSameEthAddress(from, ZERO_ADDRESS)) {
+      updatedCovTotalSupply = add(covTotalSupply, value)
+    } else if (isSameEthAddress(to, ZERO_ADDRESS)) {
+      updatedCovTotalSupply = add(covTotalSupply, value)
+    }
+
+    let arithmeticOpration = null
+    if (isSameEthAddress(address, from)) {
+      arithmeticOpration = sub
+    } else if (isSameEthAddress(address, to)) {
+      arithmeticOpration = add
+    }
```

Comment:
I think the same piece of code is placed in the `sagas/subscriptions.js` (`observeKeepTokenTransfer` method). Maybe we could actually rewrite it as a function inside `utils`? Not sure how to name it tho 🤔 Thoughts?

Reply:
Can we address it in a follow-up work? I've thought about it, but I don't have a good idea of how to deal with it in the most elegant way.